### PR TITLE
Task lifecycle memory leak fix & various cross-executor teardown scenarios.

### PR DESCRIPTION
**Pull request by vlovich**
### What does this PR do?

This fixes a memory leak GPT-6 Astra Ultra found. While the initial fix degrades performance, a subsequent optimization gets us to ~8% better than current baseline for the queued task executor micro benchmark - immediate task execution is within measurement noise. The nop benchmark is ~1-8% faster than without these changes. Task `Header` increases in size by 2 words.

There's also some clippy fixes + bumping the rust-toolchain toml to 1.98 to fix a regression opening the project in VScode (but leaving CI itself at 1.92).

After the memory leak Astra found additional shutdown scenarios where things panicked or otherwise failed. I bundled into 1 giant PR - not my jam but also I can't leverage GitHub stacked PRs since it doesn't support pushing stacks to a forked repo.

### Motivation

While reviewing #32, AI found this possible memory leak (through code inspection btw!!! crazy). This fixes that corner case. Astra kept finding additional scenarios of cross-executor cancellation leading to panics / deadlocks so I kept adding them. I separated out the tests that just improve coverage vs those that actually catch failures. All the ones that add failing tests are failing in clean main (some tests require porting).

### Related issues

#32, #40

I wonder if this is the memory leak @glommer was thinking of when he said https://github.com/DataDog/glommio/issues/448 required non-trivial ownership change fixes since this did require non-trivial ownership change fixes to fix the memory leak. The fix was definitely non-trivial.

### Additional Notes

I had AI first add a bunch of thorough testing to reproduce the failure and make sure a fix didn't break something. It added 56 tests, of which 52 were failing.

#### AI workflow

Started the effort by having it write 1 test that I hand-verified the test case was correct. Then I had it verify a bunch of additional tests that I didn't review (the other 56). Then I had it implement a fix. Then I had it reduce the Header overhead from 72 bytes to 56 bytes (net increase of 2 words, i.e. 16 bytes). Then I had it run a benchmark where I noted the substantial performance regression in micro benchmarks (~13%). Then I had it optimize performance and it ended up faster than when we started. Then I pulled out one of the independent optimizations that had nothing to do with the fix out into a separate commit.

### Checklist

[X] I have added unit tests to the code I am submitting
[X] My unit tests cover both failure and success scenarios
[] If applicable, I have discussed my architecture
[X] GitHub description authored by hand
[] Did a cursory review myself by hand

I did only a cursory scan of the AI generated tests and fixes - it seemed reasonable but it's a pretty serious non-trivial change and I don't think there's anyone left in the world that has any hands-on expertise anymore with the internals of the task management.

@humb1t we may want to add the last two things I put into the default checklist (whether the PR description was authored by hand and whether the submitter actually read the code themselves)

### CODE REVIEW

**Comment by vlovich** [comment]
Actually AI self-review caught a potential other issue. Reworking to add the new test cases and it may tweak the implementation a bit. Taking it back to draft.

**Comment by Michael-Wigham** [comment]
> possible memory leak

Seeing as this is titled `memory leak fix` can you include (or point to) the minimal test as the reproducer of this memory leak, potentially if you have some valgrind or ASAN output would be beneficial.

**Comment by vlovich** [comment]
It seems github moved this away from draft mode but I'm not done as there's still a number of regressions to fix.

But to answer your question @Michael-Wigham the minimal repro is https://github.com/glommio/glommio/pull/51/changes/715772325e5008a9f2f02e1c935004222b85ec6a#diff-848a74323fff09a0da6b79d267fe5975b2a9abd9dae455d0d7008ac6f2f72c72R16

It was expanded by AI from this baseline to check a bunch of different situations. It's possible there's a lot of unhelpful overlap added by the new test suite - I plan on having AI do another pass to check for that:
```
#[test]
fn completed_task_is_destroyed_after_executor_shutdown() {
    let baseline = TaskDebugger::task_count();
    let executor = LocalExecutor::default();

    // Complete the root task, returning its own waker.
    // run() consumes and drops the task's join handle.
    let waker = executor.run(poll_fn(|cx| {
        Poll::Ready(cx.waker().clone())
    }));

    drop(executor);
    drop(waker);

    assert_eq!(TaskDebugger::task_count(), baseline);
}
```

**Comment by vlovich** [comment]
Ok - so there's a bunch more fixes for pre-existing failures around cross-executor cancellation of resources (e.g. timers). It was kind of annoying to split out into a separate PR so it's all bundled in. I've separated commits that just improve coverage into their own commit - all other commits adding tests fail on main (I've hand-verified that).

**Comment by humb1t** [comment]
Thanks @vlovich - I think it is the right decision to keep them in one PR

**Comment by dahankzter** [comment]
This is interesting and I ran #32 and main also just to confirm your initial leak @vlovich :

| branch | result |
| ------ | ------ |
| `main` (clean) | **leaks** — payload never dropped, `left: 0, right: 1` |
| `perf/header-indices` (#32) | **leaks, identically** |
| `task-lifecycle-memory-leak` (#51) | **passes** |

**Comment by dahankzter** [comment]
This PR and #32 are fundamentally at odds regarding the header and what we store there. We should perhaps settle on that as well.

**Comment by vlovich** [comment]
I add new fields but they shouldn't conflict with your changes. What do you see as the conflict?

**Comment by humb1t** [comment]
@vlovich I think PR has conflicts with main

**Comment by dahankzter** [comment]
> I add new fields but they shouldn't conflict with your changes. What do you see as the conflict?

#32 no longer carries the notifier but I guess it not really fundamental unless your fix requires that field.

**Comment by vlovich** [comment]
@humb1t the conflict is resolved. Conflicted with #40.

**Comment by humb1t** [comment]
Thanks @vlovich - I have no tech expertise to conduct the proper review, I hope @glommio/maintainers will jump in and take a look soon 💜

**Comment by vlovich** [comment]
@humb1t AFAICT you're the only member in that group. How should we proceed?

**Comment by humb1t** [comment]
@dahankzter can you please make a tech review?

File: glommio/src/executor/context.rs [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    }
+
+    /// Provides queue and I/O context without retaining the public executor.
+    pub(super) fn with_cleanup<T>(
+        &self,
+        queue: Option<Rc<RefCell<TaskQueue>>>,
+        f: impl FnOnce() -> T,
+    ) -> T {
+        let was_active = self.tasks.cleanup_active.replace(true);
+        defer!(self.tasks.cleanup_active.set(was_active));
```
Comment (dahankzter) [inline glommio/src/executor/context.rs:69]: These various defers are a bit scary. There are two issues I think.

- It's hard to follow the flow. Something like:
```rust
struct InCleanup<'a>(PhantomData<&'a Ctx>);

struct CleanupGuard<'a> {
    ctx: &'a Ctx,
    was_active: bool,
    previous_queue: u32,
    outermost: bool,
}

impl<'a> CleanupGuard<'a> {
    fn enter(ctx: &'a Ctx, queue: u32) -> Self {
        let was_active = ctx.cleanup_active.replace(true);
        let previous_queue = ctx.active_queue.replace(queue);
        Self { ctx, was_active, previous_queue, outermost: !was_active }
    }
}

impl Drop for CleanupGuard<'_> {
    fn drop(&mut self) {
        // Everything is still installed here. That is the whole point: the
        // drain sees the queue context and the cleanup flag it requires.
        if self.outermost && self.ctx.shutting_down.get() {
            self.ctx.finish_shutdown(&InCleanup(PhantomData));
        }
        self.ctx.active_queue.set(self.previous_queue);
        self.ctx.cleanup_active.set(self.was_active);
    }
}

impl Ctx {
    fn with_cleanup<T>(&self, queue: u32, f: impl FnOnce() -> T) -> T {
        let _guard = CleanupGuard::enter(self, queue);
        f()
    }

    /// No `debug_assert`: the context is installed or this does not compile.
    fn finish_shutdown(&self, _proof: &InCleanup<'_>) { /* drain */ }
}
```
This is a small showcase to show how to get the drain before restore logic right. 

- I believe it's possible to have a panicking destructor abort since the drain runs within a Drop. Perhaps:
```rust
impl Drop for CleanupGuard<'_> {
    fn drop(&mut self) {
        if self.outermost
            && self.ctx.shutting_down.get()
            && !std::thread::panicking()   // <-- a panic is already tearing this down
        {
            self.ctx.finish_shutdown(&InCleanup(PhantomData));
        }
        ...restore...
    }
}
```
Or we can leave it since we already have abort_on_panic pattern as the idiomatic way.

File: glommio/src/task/header.rs [hunk 2]
```diff
@@ -49,32 +62,21 @@ pub(crate) struct Header {
...
-
-        // Mark the task as closed.
-        self.state |= CLOSED;
-    }
-
     /// Notifies the awaiter blocked on this task.
     ///
     /// If the awaiter is the same as the current waker, it will not be
     /// notified.
     #[inline]
```
Comment (dahankzter) [inline glommio/src/task/header.rs:76]: Generally a good idea but did you consider using `Cell` here instead of unsafe?

File: rust-toolchain.toml [hunk 3]
```diff
@@ -1,5 +1,5 @@
 [toolchain]
-channel = "1.92.0"
+channel = "1.98.1"
```
Comment (dahankzter) [inline rust-toolchain.toml:2]: Are we sure that CI stays at 1.92?

**Review by dahankzter — COMMENTED** [review]
Generally good stuff indeed. I have two comments which could be me misunderstanding things.

File: glommio/src/executor/context.rs (hunk 1 above)
Comment (dahankzter) [reply glommio/src/executor/context.rs:69]: I simply have a hard time following the defer logic and the snippet above is meant to illustrate a way to simplify it. Maybe the surrounding code makes it impossible.

File: glommio/src/executor/context.rs [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+            }
+            f()
+        })
+    }
+
+    /// Drains shutdown work once the outermost cleanup callback has returned.
+    /// Take one runnable at a time so destructors run without queue borrows and
+    /// any tasks or queues they create are included in the next iteration.
+    fn finish_shutdown(&self) {
+        loop {
```
Comment (dahankzter) [inline glommio/src/executor/context.rs:97]: What terminates this loop? 

The registry keeps accepting tasks after request_shutdown (that is what is_shutting_down is for), and a destructor that spawns is supported and tested. A future whose Drop spawns a task whose Drop spawns a task would keep this going while looking like progress. Finalizer queues usually cap it: run to a fixed point at most N times, then stop honoring new registrations and cancel instead of schedule.

Not sure if this is correct or me missing something but it looks like it's theoretically possible.

File: glommio/src/task/header.rs (hunk 2 above)
Comment (dahankzter) [reply glommio/src/task/header.rs:76]: On main both notify and register hold a &mut Header across a call into a user-supplied waker vtable, which can re-enter through a raw pointer. That is an aliasing violation under Stacked Borrows, so this fixes something the description does not claim.

A Cell removes the exclusive borrow entirely rather than avoiding it with raw pointers, so the callers stay safe at no layout cost.

**Review by dahankzter — COMMENTED** [review]
A couple of more comments.

File: rust-toolchain.toml (hunk 3 above)
Comment (vlovich) [reply rust-toolchain.toml:2]: Yes

**Comment by vlovich** [comment]
Thanks for the review. I'll work on it & hope to get an update sometime next week.

File: glommio/src/executor/context.rs (hunk 1 above)
Comment (vlovich) [reply glommio/src/executor/context.rs]: @dahankzter can you please elaborate? I feel like your proposal is much harder to follow. The defer's are written right after the site where the state was modified and needs to be modified back at the end of the scope including unwind. The custom guard I feel like is much harder to follow and is basically an unrolled state machine.

I'd prefer to leave the existing abort_on_panic pattern.

File: glommio/src/task/header.rs (hunk 2 above)
Comment (vlovich) [reply glommio/src/task/header.rs]: AFAICT making it a cell isn't really any different - Header::notify is always called from an unsafe context already so you're not going to notice that its safety changed. Additionally, you've still got aliasing concerns because making the vtable a Cell doesn't solve the other references to the other fields that are paired with the call site of notify / reentrancy concerns.

File: glommio/src/executor/context.rs (hunk 4 above)
Comment (vlovich) [reply glommio/src/executor/context.rs]: You're right that an endless chain of destructors spawning tasks can prevent shutdown from terminating. But cancelling new tasks doesn’t solve that: cancellation still drops their futures, invoking the destructors that spawn more tasks. You're only options in such a situation are leaking resources or aborting. In practice I think it's not a realistic scenario worth worrying about.

**Review by vlovich — COMMENTED** [review]
Switched to Cell as requested. I'm not convinced about the usage of the unrolled state machine vs just using defer!. To me that reads easier. AFAICT the only comment was about Cell so I believe I've answered all the questions. @dahankzter can you please confirm?

File: glommio/src/executor/context.rs (hunk 4 above)
Comment (dahankzter) [reply glommio/src/executor/context.rs]: Yeah you're right. Cancelling still drops the future so the destructors run anyway and the chain just moves off the loop and onto the stack, so a stack overflow instead of a hang. Leak or abort then like you say makes sense.

Can it be diagnosable if someone does manage to hit it? Right now it just spins with nothing in the log. Would a round counter that warns past some N be worth having? Or is that just noise?

File: glommio/src/executor/context.rs (hunk 1 above)
Comment (dahankzter) [reply glommio/src/executor/context.rs]: Maybe it's just me getting caught in Go mode by the "defer" naming :) Let's leave it as is.

File: glommio/src/executor/context.rs (hunk 4 above)
Comment (vlovich) [reply glommio/src/executor/context.rs]: I don't think it's worth it. This is just a variation of the halting problem and not some new behavior by this PR.

**Comment by vlovich** [comment]
@dahankzter I believe all comments are resolved. If so please approve so that @humb1t can merge.

**Review by dahankzter — APPROVED** [review]
